tracer/src/Datadog.Trace/HttpOverStreams/ChunkedEncodingReadStream.cs

```diff
@@ -0,0 +1,323 @@
+
+                    if (_bytesRemainingInChunk == 0)
+                    {
+                        _state = ParsingState.ExpectChunkTerminator;
+                    }
+
+                    return bytesToConsume;
+                case ParsingState.ExpectChunkTerminator:
+                    Debug.Assert(_bytesRemainingInChunk == 0, $"Expected {nameof(_bytesRemainingInChunk)} == 0, got {_bytesRemainingInChunk}");
+
+                    while (!TryReadNextChunkLine(out currentLine))
+                    {
+                        // We didn't have enough data to read the chunk terminator, so we need to refill the buffer
+                        // and try again
+                        await FillAsync().ConfigureAwait(false);
+                    }
+
+                    if (currentLine.Length != 0)
+                    {
+                        throw new Exception("Invalid response chunk terminator: expected 0 length terminator, received " + Encoding.ASCII.GetString(_streamBuffer, currentLine.Offset, currentLine.Length));
+                    }
+
+                    _state = ParsingState.ExpectChunkHeader;
+                    break;
+
+                case ParsingState.Done:
+                    // This _shouldn't_ be called once we're done, but a badly behaved
+                    // parser may call read on the stream again after it's already received
+                    // a 0 for the end of the stream
+                    return 0;
+            }
+        }
+    }
+
+    private bool TryReadNextChunkLine(out Segment chunkHeader)
+    {
+        // try to read the data we have in _currentPosition for a `\n`
+        // if we don't have anything in the buffer, then we need to do a fetch
+
+        // This will actually potentially search _beyond_ the remaining values, but just handle it later
+        var lineFeedIndex = Array.IndexOf(_streamBuffer, (byte)'\n', _currentPosition.Offset);
+        if (lineFeedIndex < 0 || lineFeedIndex >= _currentPosition.Offset + _currentPosition.Length)
+        {
+            // didn't find a line feed in the available bytes
+            if (_currentPosition.Length < MaxChunkBytesAllowed)
+            {
+                // In this case, we only have a few bytes that we've looked through
+                // So we need to extend the buffer and try again
+                chunkHeader = default;
+                return false;
+            }
+
+            // Oops, we _should_ have found it, so something is wrong here, so throw chunk size too big
+        }
+        else
+        {
+            int bytesConsumedIndex = lineFeedIndex + 1;
+            if ((bytesConsumedIndex - _currentPosition.Offset) <= MaxChunkBytesAllowed)
+            {
+                int carriageReturnIndex = lineFeedIndex - 1;
+
+                int lengthFromStart = (uint)carriageReturnIndex < (uint)(_currentPosition.Offset + _currentPosition.Length) && _streamBuffer[carriageReturnIndex] == '\r'
```

Comment:
I think carriageReturnIndex is always smaller than `_currentPosition.Offset + _currentPosition.Length` (otherwise we would be in the other branch of the if block). However, `lineFeedIndex` could be 0, in which case `_streamBuffer[carriageReturnIndex]` will read at -1 and throw

Reply:
This is basically copied from the .NET runtime, and I went spelunking, because I was a bit confused about how this could have slipped past.

[It looks like in .NET 7 this](https://github.com/dotnet/runtime/pull/69250):
```csharp
int length = buffer.IndexOf((byte)'\n');
line = buffer.Slice(0, length > 0 && buffer[length - 1] == '\r' ? length - 1 : length);
```
(which has a length > 0 check)

went to this

```
int carriageReturnIndex = lineFeedIndex - 1;

int length = (uint)carriageReturnIndex < (uint)buffer.Length && buffer[carriageReturnIndex] == '\r'
    ? carriageReturnIndex
    : lineFeedIndex;

line = buffer.Slice(0, length);
```

which has a different check, `(uint)carriageReturnIndex < (uint)buffer.Length`

It took me a moment, but that's actually a bounds check on _both_ ends, because if `carriageReturnIndex == -1`, then `(uint)carriageReturnIndex == uint.MaxValue`, so it fails the bounds check.

Not arguing that this is dangerously "clever", so going to add a comment 😄

Reply:
Oof too clever